pkg/morph/event/notary_preparator.go

```diff
@@ -0,0 +1,243 @@
+
+	// ErrTXAlreadyHandled is returned if received TX has already been signed.
+	ErrTXAlreadyHandled = errors.New("received main tx has already been handled")
+)
+
+// PreparatorPrm groups the required parameters of the Preparator constructor.
+type PreparatorPrm struct {
+	alphaKeys client.AlphabetKeys
+}
+
+// Preparator implements NotaryPreparator interface.
+type Preparator struct {
+	// contractSysCall contract call in NeoVM
+	contractSysCall []byte
+	// dummyInvocationScript is invocation script from TX that is not signed.
+	dummyInvocationScript []byte
+
+	alphaKeys client.AlphabetKeys
+}
+
+// notaryPreparator inits and returns NotaryPreparator.
+//
+// Considered to be used for preparing notary request
+// for parsing it by event.Listener.
+func notaryPreparator(prm PreparatorPrm) NotaryPreparator {
+	switch {
+	case prm.alphaKeys == nil:
+		panic("alphabet keys source must not be nil")
+	}
+
+	contractSysCall := make([]byte, 4)
+	binary.LittleEndian.PutUint32(contractSysCall, interopnames.ToID([]byte(interopnames.SystemContractCall)))
+
+	dummyInvocationScript := append([]byte{byte(opcode.PUSHDATA1), 64}, make([]byte, 64)...)
+
+	return Preparator{
+		contractSysCall:       contractSysCall,
+		dummyInvocationScript: dummyInvocationScript,
+		alphaKeys:             prm.alphaKeys,
+	}
+}
+
+// Prepare converts raw notary requests to NotaryEvent.
+//
+// Returns ErrTXAlreadyHandled if transaction shouldn't be
+// parsed and handled. It is not "error case". Every handled
+// transaction is expected to be received one more time
+// from the Notary service but already signed. This happens
+// since every notary call is a new notary request in fact.
+func (p Preparator) Prepare(nr *payload.P2PNotaryRequest) (NotaryEvent, error) {
+	currentAlphabet, err := p.alphaKeys()
+	if err != nil {
+		return nil, fmt.Errorf("could not fetch Alphabet public keys: %w", err)
+	}
+
+	// validate main TX's notary attribute
+	err = p.validateAttributes(nr.MainTransaction.Attributes, currentAlphabet)
+	if err != nil {
+		return nil, err
+	}
+
+	// validate main TX's witnesses
+	err = p.validateWitnesses(nr.MainTransaction.Scripts, currentAlphabet)
+	if err != nil {
+		return nil, err
+	}
+
+	var (
+		opCode opcode.Opcode
+		param  []byte
+	)
+
+	ctx := vm.NewContext(nr.MainTransaction.Script)
+	ops := make([]Op, 0, 16) // 16 is maximum num of opcodes for calling contracts with 4 args(no arrays of arrays)
+
+	for {
+		opCode, param, err = ctx.Next()
+		if err != nil {
+			return nil, fmt.Errorf("could not get next opcode in script: %w", err)
+		}
+
+		if opCode == opcode.RET {
+			break
+		}
+
+		ops = append(ops, Op{code: opCode, param: param})
+	}
+
+	opsLen := len(ops)
+
+	// check if it is tx with contract call
+	if !bytes.Equal(ops[opsLen-1].param, p.contractSysCall) {
+		return nil, errNotContractCall
+	}
+
+	// retrieve contract's script hash
+	contractHash, err := util.Uint160DecodeBytesBE(ops[opsLen-2].param)
```

Comment:
Don't we need to verify contract hash and method? Because now you allow *any* method of *any* contract to be called with *any* set of arguments which can be constructed from `PUSH*` opcodes. And this transaction will be successfully signed by alphabet nodes. It seemed to be quite dangerous.